sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala

```diff
@@ -106,7 +106,7 @@ case class InMemoryRelation(
 
   private def buildBuffers(): Unit = {
     val output = child.output
-    val cached = child.execute().mapPartitionsInternal { rowIterator =>
+    val cached = child.execute().mapPartitionsWithIndex { (i, rowIterator) =>
```

Comment:
Here, I'd prefer to use `TaskContext.get.getPartitionId` to get the partition id. The problem with this change is that it's re-introducing closure cleaning overhead, which `mapPartitionsInternal` avoids.